### PR TITLE
Classifying Alzheimer's Disease using the GFNet on ADNI Dataset (47438886) 

### PR DESCRIPTION
**Pull request by Kevin-Gu-2022-UQ**
Hi teaching team, 

This pull request contains the source code used to train one of the latest vision transformers, known as the GFNet, to identify Alzheimer's Disease in 2D slices of MRI brain scans. The data used comes from the [ADNI dataset](https://adni.loni.usc.edu/), which is split into two classes: AD (Alzheimer's Disease) and NC (Normal Control). An accuracy of 75.2% was achieved on the ADNI test set.

## Dependencies

While the latest dependencies will likely work, the code was written with these versions.

| Dependency | Version |
|------------|---------|
| matplotlib | 3.8.4 |
| pandas | 2.2.3 |
| Pillow | 11.0.0 |
| scikit_learn | 1.5.2 |
| timm | 1.0.11 |
| torch | 2.4.0 |
| torchvision | 0.19.0 |

### Files
The code consists of the following files:
- `modules.py`: This file contains the source code for the GFNet model. It is largely based on the code written by Rao., et al. which can be found [here](https://github.com/raoyongming/GFNet/blob/master/gfnet.py)
- `dataset.py`: This file contains the data loader for the ADNI dataset. It also performs pre-processing on the dataset.
- `train.py`: This file contains the code used to train, validate and test the GFNet model. The best model is also saved to be used in inference if desired. Relevant plots are also generated here.
- `predict.py`: This file contains the code used to run inference on the trained model on unseen test data.
- `utils.py`: This file contains utility functions useful across the multiple files. Specifically, functions to get the current device and also dataset root.
-  `requirements.txt`: This file contains the list of dependencies and their versions

Please see the `README.md` file for more information related to the algorithm, process and results.

Thank you,
Kevin Gu 47438886

### CODE REVIEW

**Comment by wangzhaomxy** [comment]
**<This is an initial inspection, no action is required at this point.>**

**File Organizing:** Well-organized files.

**Problem Solving:**

- The algorithm solves the problem appropriately. 
- Accuracy in the testing dataset: 0.752. Despite a reasonable and detailed description in the report, it still needs evidence for the testing acc, such as a screenshot of the predicted output.

**Model and functions:**

- It correctly uses PyTorch to construct the GFNet models and functions.
- With data augmentation.
- The output of the predict.py is expected to be a classification label with an image input instead of printing an accuracy number.

**Code design:** Good.

**Code comment and docstring:**

- Good code comments 
- Good function docstrings

**Difficulty:** Hard.

**Additional Comments:**

- intermediate numbers of commits
- Good ReadMe design and fruitful content.

**Comment by aniketgupta17** [comment]
Observational Feedback

Pull Request:
Correctly created the Pull request from Topic Recognition Branch .
The pull request includes a clear description about the file structure .

File Organizing: Well-organized files.

Commit Log:
Commit messages are progressive for the Recognition Problem solved using 4 files . 
Commits are regularly made, showing incremental development rather than bulk updates.

Documentation:
Code comments and docstrings  are included.
Proper GitHub markdown formatting is used, with organized headings, lists, and code blocks.
Can add about Conclusion and  future improvements in the ReadMe.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. The output of the predict.py is expected to be a classification</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Uses Data Aug.</td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  GFNet (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Accuracy information missing.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by Kevin-Gu-2022** [comment]
Hi Gayan,

I was just wondering what this line meant.

> Readme :Acceptable. Accuracy information missing.	-1

I do have the test accuracy at the bottom of my README.

**Comment by wangzhaomxy** [comment]
> Hi Gayan,
> 
> I was just wondering what this line meant.
> 
> > Readme :Acceptable. Accuracy information missing.	-1
> 
> I do have the test accuracy at the bottom of my README.

Hi Kevin,

I believe this drop-point is based on my previous comment: "Accuracy in the testing dataset: 0.752. Despite a reasonable and detailed description in the report, it still needs evidence for the testing acc, such as a screenshot of the predicted output."

Cheers,
Zhao

**Comment by Kevin-Gu-2022-UQ** [comment]
> Hi Kevin,
> 
> I believe this drop-point is based on my previous comment: "Accuracy in the testing dataset: 0.752. Despite a reasonable and detailed description in the report, it still needs evidence for the testing acc, such as a screenshot of the predicted output."
> 
> Cheers, Zhao

Hi Zhao, 

If I fix this error now, would it mean I can get that mark back? 

Cheers,
Kevin

**Comment by wangzhaomxy** [comment]
> > Hi Kevin,
> > I believe this drop-point is based on my previous comment: "Accuracy in the testing dataset: 0.752. Despite a reasonable and detailed description in the report, it still needs evidence for the testing acc, such as a screenshot of the predicted output."
> > Cheers, Zhao
> 
> Hi Zhao,
> 
> If I fix this error now, would it mean I can get that mark back?
> 
> Cheers, Kevin

Hi Kevin,

Unfortunately, you cannot get this mark back. Please see the second line from the bottom in Gayan's marking comment. We only accept changes before the due date other than what the marking feedback tells you to do.
